### PR TITLE
CPP/SQL: Fix Mining/Gathering while Mounted

### LINKED ISSUE #197: Ores can be mined from the mount
I am unsure that is supposed to happen, because otherwise mounts like sky golem wouldn't exist
Image

### PR DESCRIPTION
**Pull request by CDBrodie**
## Changes Proposed:

This PR proposes changes to:
-  [X] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [X] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes #197 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [X] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR fixes issue #197 where players were able to mine ore deposits while mounted without having an effect that should permit mounted gathering.

### Issue

Players were able to successfully mine ore deposits while mounted on a normal mount.

This was not limited to the interaction cursor or a client-side visual issue. The server allowed the mining interaction to complete, the gathering cast succeeded, and the player could loot the mining node without dismounting.

The expected behavior is:

- Normal mounted players should not be able to mine.
- Unmounted players should be able to mine normally.
- Players with an effect specifically allowing mounted gathering, such as Monel-Hardened Stirrups, should be able to mine while mounted.

### Root Cause

Mining ultimately uses the `SPELL_EFFECT_OPEN_LOCK` path.

HavenCore already performs generic mounted spell validation, but this was not sufficient for gathering interactions because the gathering-related spell can pass the generic mounted-cast checks.

The `SPELL_EFFECT_OPEN_LOCK` validation checked the target, lock information, and skill requirements, but did not perform an additional mounted-use check against the targeted gathering gameobject.

As a result, the gathering interaction could proceed while the player was mounted.

This affected gathering objects using both:

- `GAMEOBJECT_TYPE_GATHERING_NODE`
- `GAMEOBJECT_TYPE_CHEST`

### Mounted Gathering Fix

Additional validation was added to the `SPELL_EFFECT_OPEN_LOCK` handling in:

`src/server/game/Spells/Spell.cpp`

When the target is a gathering node or chest, the server now checks whether the player is mounted before allowing the interaction to proceed.

The gathering cast is rejected when all of the following are true:

- The target is `GAMEOBJECT_TYPE_GATHERING_NODE` or `GAMEOBJECT_TYPE_CHEST`.
- The player is mounted.
- The gameobject is not explicitly configured as usable while mounted.
- The player does not have `PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED`.

When these conditions are met, the server returns:

`SPELL_FAILED_NOT_MOUNTED`

This prevents normal mounted mining while retaining support for gameobjects and player effects that legitimately permit mounted interaction.

### Monel-Hardened Stirrups

After normal mounted mining was blocked, an additional issue was discovered with Monel-Hardened Stirrups.

Monel-Hardened Stirrups are intended to permit gathering while mounted, but the mounted gathering validation initially blocked mining even while the Stirrups effect was active.

The active spell was identified as:

`267560 - Monel-Hardened Stirrups`

The loaded `SpellEffect.db2` data was inspected directly and returned:

- Spell: `267560`
- Effect: `6`
- Aura: `4`
- Aura type: `SPELL_AURA_DUMMY`

HavenCore already has support for:

`SPELL_AURA_ALLOW_USING_GAMEOBJECTS_WHILE_MOUNTED`

which manages:

`PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED`

However, spell `267560` does not use that aura type in the BFA data. It is implemented as `SPELL_AURA_DUMMY`.

Therefore, applying Monel-Hardened Stirrups never caused `PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED` to be set.

### Monel-Hardened Stirrups Fix

A dedicated aura script was added to:

`src/server/scripts/Spells/spell_generic.cpp`

for:

`267560 - Monel-Hardened Stirrups`

When the aura is applied, the script adds:

`PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED`

When the aura is removed or expires, the script removes:

`PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED`

This allows the existing mounted-object permission system to be used rather than hardcoding an exception for spell `267560` directly into the gathering logic.

The `Spell.cpp` mounted gathering check therefore remains generic and can recognize any effect that correctly grants the mounted gameobject-use permission.

### Database Change

A world database update was added to register spell `267560` with:

`spell_gen_monel_hardened_stirrups`

in `spell_script_names`.

This causes the new aura script to execute when Monel-Hardened Stirrups are applied or removed.

### Result

The resulting behavior is:

- Unmounted + normal mining node: mining works.
- Mounted + normal mining node: mining is blocked.
- Mounted + Monel-Hardened Stirrups: mining works.
- Mounted after Stirrups expire/remove: mining is blocked again.
- Objects explicitly configured as usable while mounted remain usable while mounted.

### Investigation Performed

The mounted gathering interaction was traced through several parts of the core before the final fix was implemented.

The following areas were investigated:

- `HandleGameObjectUseOpcode()`
- `Spell::CheckCast()`
- `SPELL_EFFECT_OPEN_LOCK`
- `EffectOpenLock()`
- `CanOpenLock()`
- generic mounted spell validation
- `IsMounted()`
- `SPELL_AURA_ALLOW_USING_GAMEOBJECTS_WHILE_MOUNTED`
- `PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED`
- `SpellEffect.db2`
- Monel-Hardened Stirrups spell `267560`

Temporary diagnostic logging was used to inspect the loaded `SpellEffect.db2` information for Monel-Hardened Stirrups.

The diagnostic output confirmed:

`Spell=267560 EffectID=714314 Index=0 Effect=6 Aura=4`

All temporary diagnostic/debugging code was removed from the final changes.

### Build Testing

The final changes were compiled successfully using:

- Windows
- x64
- RelWithDebInfo

The scripts project compiled successfully and generated the expected `scripts.lib`.

No build errors were introduced by the final changes.

#### Test 1 - Normal Mining While Unmounted

1. Logged into a character with Mining.
2. Located a valid mining node.
3. Attempted to mine while unmounted.

Result:

- Mining completed normally.
- The node was gathered.
- Loot was available normally.

#### Test 2 - Normal Mining While Mounted

1. Mounted on a normal mount.
2. Located a valid mining node.
3. Attempted to mine without any mounted-gathering effect.

Result:

- Mining was blocked.
- The gathering interaction did not complete.
- The mining node was not consumed.

#### Test 3 - Dismount and Mine

1. Attempted to mine while mounted and confirmed it was blocked.
2. Dismounted.
3. Attempted to mine the same type of node.

Result:

- Mining completed normally after dismounting.

#### Test 4 - Monel-Hardened Stirrups

1. Applied Monel-Hardened Stirrups.
2. Confirmed spell `267560` was active.
3. Mounted normally.
4. Located a valid mining node.
5. Attempted to mine while mounted.

Result:

- Mining completed successfully while mounted.
- The node was gathered.
- Loot was available normally.

#### Test 5 - Mounted Mining Without Stirrups

1. Tested the same mounted mining behavior without Monel-Hardened Stirrups active.

Result:

- Mining was blocked as expected.

#### Test 6 - BFA Gathering Nodes

BFA `GAMEOBJECT_TYPE_GATHERING_NODE` mining objects were tested.

Result:

- Normal mounted mining was blocked.
- Unmounted mining worked.
- Mining while mounted with Monel-Hardened Stirrups worked.

This PR has been:

- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Apply the world database update included with the PR.

2. Confirm that spell `267560` is registered in `spell_script_names`.

   Expected ScriptName:

   `spell_gen_monel_hardened_stirrups`

3. Rebuild the server using the updated `Spell.cpp` and `spell_generic.cpp`.

4. Restart worldserver.

5. Log into a character with Mining.

6. Ensure the character has sufficient Mining skill for the node being tested.

7. Locate a valid mining node.

8. While unmounted, attempt to mine the node.

   Expected:

   - Mining succeeds.
   - Loot is generated normally.

9. Locate another valid mining node.

10. Mount a normal ground or flying mount.

11. Ensure Monel-Hardened Stirrups are NOT active.

12. Attempt to mine while mounted.

   Expected:

   - Mining is rejected.
   - The player cannot gather the node while mounted.
   - The node is not consumed.

13. Dismount.

14. Attempt to mine again.

   Expected:

   - Mining succeeds normally.

15. Apply Monel-Hardened Stirrups.

16. Confirm aura `267560` is active.

17. Mount normally.

18. Locate another valid mining node.

19. Attempt to mine while mounted.

   Expected:

   - Mining succeeds while mounted.
   - The gathering interaction completes.
   - Loot is generated normally.

20. Remove Monel-Hardened Stirrups or allow the aura to expire.

21. Remain mounted.

22. Attempt to mine another valid node.

   Expected:

   - Mining is blocked again.

23. Test a BFA mining node using `GAMEOBJECT_TYPE_GATHERING_NODE`.

   Expected:

   - Unmounted mining works.
   - Normal mounted mining is blocked.
   - Mounted mining with Monel-Hardened Stirrups works.

24. Test an older mining node using `GAMEOBJECT_TYPE_CHEST`.

   Expected:

   - Unmounted mining works.
   - Normal mounted mining is blocked.
   - Mounted mining with the appropriate mounted-use permission remains possible.

25. Test a gameobject that is explicitly configured as usable while mounted.

   Expected:

   - The new mounted gathering validation does not incorrectly block objects that explicitly permit mounted interaction.

26. Verify normal Mining skill requirements still function.

   Attempt to gather a node for which the character does not meet the required Mining skill.

   Expected:

   - The normal skill requirement prevents gathering.
   - The mounted-gathering changes do not bypass `CanOpenLock()` or profession skill requirements.

27. Additional community testing is recommended for:

   - Classic/older-world mining nodes.
   - Cataclysm mining nodes.
   - Mists of Pandaria mining nodes.
   - Warlords of Draenor mining nodes.
   - Legion mining nodes.
   - Battle for Azeroth mining nodes.
   - Ground mounts.
   - Flying mounts.
   - Monel-Hardened Stirrups application and expiration.
   - Gameobjects explicitly configured for mounted interaction.

### CODE REVIEW

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -5590,6 +5590,20 @@ SpellCastResult Spell::CheckCast(bool strict, uint32* param1 /*= nullptr*/, uint
...
+{
+    GameObjectTemplate const* goInfo = go->GetGOInfo();
+
+    if ((goInfo->type == GAMEOBJECT_TYPE_GATHERING_NODE ||
+         goInfo->type == GAMEOBJECT_TYPE_CHEST) &&
+        m_caster->IsMounted() &&
+        !goInfo->IsUsableMounted() &&
+        !m_caster->ToPlayer()->HasPlayerLocalFlag(PLAYER_LOCAL_FLAG_CAN_USE_OBJECTS_MOUNTED))
+    {
+        return SPELL_FAILED_NOT_MOUNTED;
```
Comment (Hextv) [inline src/server/game/Spells/Spell.cpp]: Isn't that the wrong error? As far i understand SPELL_FAILED_NOT_MOUNTED tells the client that the player needs to be mounted or i'm wrong?

File: src/server/scripts/Spells/spell_generic.cpp [hunk 2]
```diff
@@ -1605,6 +1605,38 @@ class spell_gen_elune_candle : public SpellScriptLoader
         }
 };
```
Comment (Hextv) [inline src/server/scripts/Spells/spell_generic.cpp:1607]: A short comment here would be nice

File: src/server/game/Spells/Spell.cpp [hunk 3]
```diff
@@ -205,7 +205,7 @@ void SpellCastTargets::Write(WorldPackets::Spells::SpellTargetData& data)
 
     if (m_targetMask & TARGET_FLAG_SOURCE_LOCATION)
     {
-        data.SrcLocation.emplace();
+        data.SrcLocation = boost::in_place();
```
Comment (Xenphis) [inline src/server/game/Spells/Spell.cpp]: update your branch. You will revert some recent changes.

**Review by Xenphis — CHANGES_REQUESTED** [review]
